### PR TITLE
Add Truncate flag to fs_open - Implementation and unit tests

### PR DESCRIPTION
**Pull request by RAJAGOPALAN-GANGADHARAN**
### Introduction:
------------------

This PR is supporting the implementations of posix functions like fread (https://github.com/zephyrproject-rtos/zephyr/issues/66940), fopen.

According to ISO-C standard, 'w*' mode passed to fopen should create a new file or else truncate the contents of the existing file. While implementing fopen, I found that it was a bit unclean to convert posix flags into zfs flags. ZFS supports posix equivalents for e.g: 'r' => FS_O_READ. It would be easier to implement ISO-C's fopen behavior if ZFS natively supports truncate flag and truncates during the file open stage.

------------------
### Considerations:
------------------

I considered two approaches here:

1. Create a flag for truncate in fs and make changes to fs_open method to also call truncate when this flag is passed (its possible that an underlying fs might have not yet implemented the truncate method)
(Or)
2. Directly handle truncation at fopen level. On a side note we can have a flag for truncate like FS_O_TRUNC as well but this would only make sense if option 1 is picked.

The first option made sense to me as there can be client usecases where a file needs to be truncated during opening regardless of the fs. So after having a chat with @cfriedt , making this PR without having an issue attached.

Other Considerations:
1. Write flag has to be passed alongside truncate, otherwise the open operation fails with Access denied.
2. If truncate operation is not defined for an fs, it fails with Operation unsupported error with a log. Maybe this can be handled in a better way by just recreating the file? 

------------------
### Implementation
------------------

1. Added support for new flag.
2. Added truncate operation within fs_open based on flags.
3. Added unit tests that run flag related tests on common fs.
4. Clang formatted the modified segments (Although when I ran it on the file, I could see a lot more changes throughout the file - maybe a refactor PR is needed?)

------------------
### Testing
------------------

Ran filesystem related tests which is relevant to this patch:

```
./scripts/twister -i -T tests/subsys/fs --tag filesystem
```

No failures found. Relying on CI for extensive testing.

### CODE REVIEW

File: include/zephyr/fs/fs.h [hunk 1]
```diff
@@ -162,6 +162,8 @@ struct fs_statvfs {
 #define FS_O_APPEND     0x20
 /** Bitmask for open/create flags */
 #define FS_O_FLAGS_MASK 0x30
+/** Truncate the file while opening */
+#define FS_O_TRUNC      0x40
```
Comment (de-nordic) [inline include/zephyr/fs/fs.h:166]: This should be after `FS_O_APPEND`, you also need to fix the `FS_O_FLAGS_MASK` that does not cover the flag.

File: subsys/fs/fs.c [hunk 2]
```diff
@@ -168,6 +168,25 @@ int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags)
 		return rc;
 	}
 
+	if (flags & FS_O_TRUNC) {
+		if ((flags & FS_O_WRITE) == 0) {
```
Comment (de-nordic) [inline subsys/fs/fs.c:172]: This check should happen before open is even attempted. Yo do not perform any operation if it is possible to predict failure.

File: subsys/fs/fs.c [hunk 3]
```diff
@@ -168,6 +168,25 @@ int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags)
...
+			/** Truncate not allowed when file is not opened for write */
+			LOG_ERR("file should be opened for write to truncate!!");
+			return -EACCES;
+		}
+		CHECKIF(mp->fs->truncate == NULL) {
+			LOG_ERR("file truncation not supported!!");
+			return -ENOTSUP;
+		}
```
Comment (de-nordic) [inline subsys/fs/fs.c:180]: This part should be directly tested with the check for `FS_O_TRUNC`, fail if you can predict failure before doing anything.

**Review by de-nordic — CHANGES_REQUESTED** [review]
Comments, otherwise I think it is fine.

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
Sure thanks for reviewing @de-nordic , will address the comments and ci failures :D

File: subsys/fs/fs.c (hunk 3 above)
Comment (RAJAGOPALAN-GANGADHARAN) [reply subsys/fs/fs.c:180]: A good catch!

**Review by cfriedt — DISMISSED** [review]

**Comment by de-nordic** [comment]
There are still compliance issues.
While fixing these can you also add one more commit adding information on addition of truncation flag support to fs section of release notes for 3.7?

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
> There are still compliance issues. While fixing these can you also add one more commit adding information on addition of truncation flag support to fs section of release notes for 3.7?

Done! I was not able to test the release notes documentation build locally because of resource constraints in wsl's virtual disk (keeps filling my disk for no reason). Let me check the built documentation generated from CI. Also addressed the Compliance check (but the error seems to be pointing to older commit that is now overwritten).

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
```
7eb6891c6f5e90966b244ace08c07056b735d02a: author email (RAJAGOPALAN-GANGADHARAN <g.raju2000@gmail.com>) needs to match one of the signed-off-by entries.
7eb6891c6f5e90966b244ace08c07056b735d02a: author email (RAJAGOPALAN-GANGADHARAN <g.raju2000@gmail.com>) does not follow the syntax: First Last <email>.
 22c2c9deab92bf0d390340a9f461c82be6582665: author email (RAJAGOPALAN-GANGADHARAN <g.raju2000@gmail.com>) needs to match one of the signed-off-by entries.
22c2c9deab92bf0d390340a9f461c82be6582665: author email (RAJAGOPALAN-GANGADHARAN <g.raju2000@gmail.com>) does not follow the syntax: First Last <email>.
```

while both of my commits have the following signature:

```
Signed-off-by: RAJAGOPALAN GANGADHARAN <g.raju2000@gmail.com>
```

Any idea how to fix the compliance failures?

**Comment by pdgendt** [comment]
> ```
> 7eb6891c6f5e90966b244ace08c07056b735d02a: author email (RAJAGOPALAN-GANGADHARAN <g.raju2000@gmail.com>) needs to match one of the signed-off-by entries.
> 7eb6891c6f5e90966b244ace08c07056b735d02a: author email (RAJAGOPALAN-GANGADHARAN <g.raju2000@gmail.com>) does not follow the syntax: First Last <email>.
>  22c2c9deab92bf0d390340a9f461c82be6582665: author email (RAJAGOPALAN-GANGADHARAN <g.raju2000@gmail.com>) needs to match one of the signed-off-by entries.
> 22c2c9deab92bf0d390340a9f461c82be6582665: author email (RAJAGOPALAN-GANGADHARAN <g.raju2000@gmail.com>) does not follow the syntax: First Last <email>.
> ```
> 
> while both of my commits have the following signature:
> 
> ```
> Signed-off-by: RAJAGOPALAN GANGADHARAN <g.raju2000@gmail.com>
> ```
> 
> Any idea how to fix the compliance failures?

I think you have to remove the dash in your name.

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
> I think you have to remove the dash in your name.

You mean I should change my name on github?

**Comment by kartben** [comment]
> > I think you have to remove the dash in your name.
> 
> You mean I should change my name on github? 

You should amend your author name in your commit

**Comment by pdgendt** [comment]
> As you can see the sign-off in commit is in the specified format!

It's the author name in your git config that has a dash and needs fixing.

**Review by cfriedt — DISMISSED** [review]

File: subsys/fs/fs.c [hunk 4]
```diff
@@ -160,6 +160,21 @@ int fs_open(struct fs_file_t *zfp, const char *file_name, fs_mode_t flags)
 		return -ENOTSUP;
 	}
 
+	bool truncate_file = false;
```
Comment (pdgendt) [inline subsys/fs/fs.c:163]: Nit: move this to the top of the function.

**Review by pdgendt — DISMISSED** [review]

File: subsys/fs/fs.c (hunk 4 above)
Comment (RAJAGOPALAN-GANGADHARAN) [reply subsys/fs/fs.c]: Done.

**Review by pdgendt — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]
Looks OK. Thanks.

**Pull request merged**